packages/xforms-engine/src/instance/RankControl.ts

```diff
@@ -0,0 +1,198 @@
+		const codec = new MultipleValueItemCodec(sharedValueCodecs.string);
+		super(parent, definition, codec);
+
+		this.appearances = definition.bodyElement.appearances;
+
+		const valueOptions = createItemCollection(this);
+		const mapOptionsByValue: Accessor<RankItemMap> = this.scope.runTask(() => {
+			return createMemo(() => {
+				return new Map(valueOptions().map((item) => [item.value, item]));
+			});
+		});
+
+		this.mapOptionsByValue = mapOptionsByValue;
+
+		const baseValueState = this.valueState;
+		const [baseGetValue, setValue] = baseValueState;
+
+		/**
+		 * @ToDo As new value options become available, they're not yet in the
+		 * `currentValues` state. This appends them. We intend to change this
+		 * behavior, likely clearing the previous state instead.
+		 *
+		 * However, there's an open question about what we should do when a filter
+		 * change **only removes values**.
+		 */
+		const getValue = this.scope.runTask(() => {
+			return createMemo(() => {
+				const options = valueOptions();
+				const values = baseGetValue();
+
+				if (isBlankValueState(values)) {
+					return values;
+				}
+
+				const optionValues = new Set(options.map((option) => option.value));
+
+				/**
+				 * @see {@link getValue} ToDo paragraph 2.
+				 */
+				const currentValues = values.filter((value) => optionValues.has(value));
+
+				return Array.from(
+					new Set([
+						...currentValues,
+						/**
+						 * @see {@link getValue} ToDo paragraph 1.
+						 */
+						...optionValues,
+					])
+				);
+			});
+		});
+		const valueState: SimpleAtomicState<readonly string[]> = [getValue, setValue];
+
+		this.getInstanceValue = this.scope.runTask(() => {
+			return createMemo(() => codec.encodeValue(getValue()));
+		});
+
+		const sharedStateOptions = {
+			clientStateFactory: this.engineConfig.stateFactory,
+		};
+
+		const state = createSharedNodeState(
+			this.scope,
+			{
+				reference: this.contextReference,
+				readonly: this.isReadonly,
+				relevant: this.isRelevant,
+				required: this.isRequired,
+				label: createNodeLabel(this, definition),
+				hint: createFieldHint(this, definition),
+				children: null,
+				valueOptions,
+				value: valueState,
+				instanceValue: this.getInstanceValue,
+			},
+			sharedStateOptions
+		);
+
+		this.state = state;
+		this.engineState = state.engineState;
+		this.currentState = state.currentState;
+	}
+
+	getValueLabel(value: string): TextRange<'item-label'> | null {
+		const valueOption = this.currentState.valueOptions.find((item) => item.value === value);
+		return valueOption?.label ?? null;
+	}
+
+	setValues(valuesInOrder: readonly string[]): Root {
+		if (isBlankValueState(valuesInOrder)) {
+			this.setValueState(valuesInOrder);
+			return this.root;
+		}
+
+		const sourceValues: string[] = Array.from(this.mapOptionsByValue().keys());
+		const hasAllValues = sourceValues.some((sourceValue) => valuesInOrder.includes(sourceValue));
```

Comment:
```suggestion
		const hasAllValues = sourceValues.every((sourceValue) => valuesInOrder.includes(sourceValue));
```

Once the [assertion for this check](https://github.com/getodk/web-forms/pull/288#discussion_r1937726138) is added, it'll fail before this change and pass after.

Reply:
Done